### ISSUE TITLE
Salt is seeing incorrect values when checking against source_hash

### ISSUE THREAD

**Issue by justindesilets** [saltstack/salt#44112]
### Description of Issue/Question
I can confirm this affects both sha256 files hashes and sha512 hashes. I just upgraded all of my minions and master to 2017.7.2 when the issue started. I have verified the hashes on the command line but Salt seems to see another value when it checks the file hash during the state run.
### Setup
Current state:

```
install_prometheus_node_exporter:
  file.managed:
    - name: /opt/node_exporter/node_exporter
    - source: http://repo.test.net/packages/prometheus/node_exporter/node_exporter
    - source_hash: 8aa6be677870c8b7dc3c5e1ff5824838c005b5413e2ea7781bd8763ca14c6aba04a0cad4698fc613ef013797821c39405e1566855f0c63495ebb4a2f733f75a7
    - mode: 755
    - makedirs: True
```

If I check the sha512sum of the file on that server I see:

```
[root@repo node_exporter]# sha512sum node_exporter
8aa6be677870c8b7dc3c5e1ff5824838c005b5413e2ea7781bd8763ca14c6aba04a0cad4698fc613ef013797821c39405e1566855f0c63495ebb4a2f733f75a7  node_exporter
```

### Steps to Reproduce Issue
If I run the state, I consistently receive the following error:

```
----------
          ID: install_prometheus_node_exporter
    Function: file.managed
        Name: /opt/node_exporter/node_exporter
      Result: False
     Comment: Specified sha512 checksum for /opt/node_exporter/node_exporter (8aa6be677870c8b7dc3c5e1ff5824838c005b5413e2ea7781bd8763ca14c6aba04a0cad4698fc613ef013797821c39405e1566855f0c63495ebb4a2f733f75a7) does not match actual checksum (cf83e1357eefb8bdf1542850d66d8007d620e4050b5715dc83f4a921d36ce9ce47d0d13c5d85f2b0ff8318d2877eec2f63b931bd47417a81a538327af927da3e)
     Started: 17:53:00.021319
    Duration: 48.178 ms
     Changes:
              ----------
              diff:
                  New file
----------
```

### Versions Report
Saltmaster:

```
[root@saltmaster ~]# salt --versions-report
Salt Version:
           Salt: 2017.7.2

Dependency Versions:
           cffi: 1.6.0
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: 0.6.4
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: 0.24.6
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: 2.14
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: 0.24.2
         Python: 2.7.5 (default, Aug  4 2017, 00:39:18)
   python-gnupg: 0.3.8
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: centos 7.4.1708 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-693.2.2.el7.x86_64
         system: Linux
        version: CentOS Linux 7.4.1708 Core
```

Saltminion:

```
[root@repo node_exporter]# salt-call --versions-report
Salt Version:
           Salt: 2017.7.2

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.5 (default, Aug  4 2017, 00:39:18)
   python-gnupg: Not Installed
         PyYAML: 3.10
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: centos 7.4.1708 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-693.2.2.el7.x86_64
         system: Linux
        version: CentOS Linux 7.4.1708 Core
```

**Comment by Ch3LL** [comment]
i'm not able to replicate this using this state file:

```
install_prometheus_node_exporter:   
  file.managed:                     
    - name: /tmp/test               
    - source: http://repo.saltstack.com/yum/redhat/7/x86_64/latest/python-futures-3.0.3-1.el7.noarch.rpm                                        
    - source_hash: f624d5cd53138b8092b6fbcf934fb8ed5f484cd93a10c1fb54b123d3cd53350df0a5f91137abe9d701ad644ceb8acff73bf0cf18d91673350f0a09c3f7c2f0f4
    - mode: 755                     
    - makedirs: True
```

If you try that state file do you still see the error? Aslo what version did you upgrade from?

**Comment by justindesilets** [comment]
I will test when I get home tonight. I am just running this on a little lab at home and had upgraded the master and minions from 2017.7.1 to 2017.7.2. After the upgrade, the state started to fail. I will try and dig into it a little deeper on my side.

**Comment by justindesilets** [comment]
Well, copy and pasted your state into a test.sls state and it ran clean. I have double check the sum on the host server and then I also used a wget to pull it down to one of the minions and verified that the sha512sum returned the hash that is in the state. How or where is the hash calculated? Is it done by the master or the minion before it is dropped and what is the command that is checking the hash? I don't know if it would have any bearing in this but all of these CentOS 7.4 instances are running on a VMWare box. I realize this problem now very much might be limited to just me but I would like to trace this out and find out why.

**Comment by Ch3LL** [comment]
I think you will want to debug in the `get_source_sum` method in the file execution module, but @terminalmage is more knowledgeable in this area and might be able to help trace this a bit more.

**Comment by Ch3LL** [comment]
I just tested with this state:

```
test:
  archive.extracted:
    - name: /srv/salt/node
    - source: https://github.com/prometheus/node_exporter/releases/download/v0.15.0/node_exporter-0.15.0.linux-amd64.tar.gz
    - source_hash: ad77889868bf6042108a036744cd1558bbd62ed56297b61f01053a4b48272b0af6fdb5a70969c11e64c4a237f9a3da60b05ad2013e04480aefed8cbf63a306d7

test2:
  file.managed:
    - name: /tmp/test
    - source: salt://node/node_exporter-0.15.0.linux-amd64/node_exporter
    - source_hash: db8ccf41315510a3f285d619226c12267aefb4e40456638360e022b628fb1048e2842f09fe4a4ecb07fd8ba0f8468ae6303aec30b803e1249a30ddc093da9ea4

```

Does that look accurate as to what you are trying to do?

My other thought is you stated you upgraded your master/minions. What if you deleted cache? does that help?

**Comment by justindesilets** [comment]
I did perform the upgrade from 2017.7.1 to 2017.7.2. I will try and dump the cache on the master / minions tonight and will update with the results.

**Comment by justindesilets** [comment]
I have tried clearing the cache on all nodes. I have run.

```
salt '*' saltutil.clear_cache
```

I then downloaded the latest version of the node exporter and setup the following state after using

```
sha512sum node_exporter
db8ccf41315510a3f285d619226c12267aefb4e40456638360e022b628fb1048e2842f09fe4a4ecb07fd8ba0f8468ae6303aec30b803e1249a30ddc093da9ea4  node_exporter
```

I built this state as an updated test

```
install_prometheus_node_exporter:
  file.managed:
    - name: /tmp/test
    - source: http://kickstart2.j2test.net/packages/prometheus/node_exporter-0.15.0.linux-amd64/node_exporter
    - source_hash: db8ccf41315510a3f285d619226c12267aefb4e40456638360e022b628fb1048e2842f09fe4a4ecb07fd8ba0f8468ae6303aec30b803e1249a30ddc093da9ea4
    - mode: 755
    - makedirs: True
```

Unfortunately I'm getting the same failed result on every node. This includes the node I checked the checksum on.

```
----------
          ID: install_prometheus_node_exporter
    Function: file.managed
        Name: /tmp/test
      Result: False
     Comment: Specified sha512 checksum for http://kickstart2.j2test.net/packages/prometheus/node_exporter-0.15.0.linux-amd64/node_exporter (db8ccf41315510a3f285d619226c12267aefb4e40456638360e022b628fb1048e2842f09fe4a4ecb07fd8ba0f8468ae6303aec30b803e1249a30ddc093da9ea4) does not match actual checksum (cf83e1357eefb8bdf1542850d66d8007d620e4050b5715dc83f4a921d36ce9ce47d0d13c5d85f2b0ff8318d2877eec2f63b931bd47417a81a538327af927da3e). If the 'source_hash' value refers to a remote file with multiple possible matches, then it may be necessary to set 'source_hash_name'.
     Started: 22:24:25.359634
    Duration: 105.784 ms
     Changes:
```

These are all very vanilla CentOS 7.4 installs. I cannot think of anything that I might have done on them to cause the checksum mismatch. I have tried running trace logging on both the master and minion but I never see when/where the chechsum is being calculated. Only the line with the failure about the mismatch. If there is a way to directly test the checksum function from salt I would be happy to try.

**Comment by Ch3LL** [comment]
I think i just replicated it! The key was using the http source instead of salt://. I'm going to do a git bisect to see what broke it and paste here.

**Comment by Ch3LL** [comment]
I am having a hard time getting an exact commit when trying to git bisect but when troubleshooting I found that when we pull down the file to the cache if i manually run `sha512 <filename>` it gives the same checksum that salt is using so I think the issue is related to HOW we are pulling down the file, and not how we are checking the sha sum.

I even tried a text file i uploaded to my local webserver and I got the same error, so I don't think its related to the type of file it is, I think its related to the source and the httpd server.

If I use the following state file:

```
test2:
  file.managed:
    - name: /tmp/test
    - source: http://172.17.0.3:80/node/node_exporter-0.15.0.linux-amd64/node_exporter
    - source_hash: db8ccf41315510a3f285d619226c12267aefb4e40456638360e022b628fb1048e2842f09fe4a4ecb07fd8ba0f8468ae6303aec30b803e1249a30ddc093da9ea4

test3:
  file.managed:
    - name: /tmp/test2
    - source: http://172.17.0.3:80/auth.sls
    - source_hash: 43ad3d87a212b8cc4b2202f6a36eac5d9f4cbc7f86d0b4e66961977cd38616fe837276a8612d1e5fe5c558ec4f46668e2db89904dc3f8a4dcd96bd97668783ce

test4:
  file.managed:
    - name: /tmp/test4
    - source: http://mirror.cc.columbia.edu/pub/software/apache//httpd/httpd-2.4.29.tar.bz2
    - source_hash: 840982fd3d21463bc5c1747f211dfacf3abdf634d149e49bb49fd2e5deda140de602dbdf31e1bbe5337a48f718ab2261c408e83a8dd39a9291ee7b6b7a85639a
```

The first two sources fail which are from a local web server I used just by running the standard http docker container. The other source is an external source from an apache mirror. So i'm wondering if there is something related to how these httpd servers are setup

ping @terminalmage any chance you can take a look here?

**Comment by justindesilets** [comment]
@Ch3LL Awesome work, thank you. I can tell you that this is running a very plain version of apache on my side. For my use case it was the standard package that was installed when I installed cobbler. From there I just created directories under the `/var/www/html` folder and have been pulling them from there.

**Comment by clinta** [comment]
I'm seeing this as well. And with the same wrong checksum reported from salt. cf83e1357eefb8bdf1542850d66d8007d620e4050b5715dc83f4a921d36ce9ce47d0d13c5d85f2b0ff8318d2877eec2f63b931bd47417a81a538327af927da3e

I did some testing and found that is the checksum of an empty file.

**Comment by clinta** [comment]
I wondered if this is possibly related to #44013 since that issue was also a downloaded file being empty. I tested with fileclient.py from #44016 and I confirmed that does fix this issue.

**Comment by Ch3LL** [comment]
Thanks @clinta for finding that :)  I was also able to verify that the head of 2017.7 is working for me as well. This will be included in the 2017.7.3 release.
